### ISSUE TITLE
Add Feature: Auto-scaling the UI  (on Linux)

### ISSUE THREAD

**Issue by ShalokShalom** [defold/defold#9924]
I like to integrate automated UI scaling and this is how far I got:

```clojure
(defn editor-scale []
  (let [screen (Screen/getPrimary)
        bounds (.getBounds screen)
        width (.getWidth bounds)]
    (cond
      (>= width 3840) 2.0  
      (>= width 2560) 1.5  
      :else 1.0)))         
```
I am still very new to the codebase, Clojure and programming at all. 

The editor already scales well when I set `vmargs = -Dglass.gtk.uiScale=200%`, but I like it to auto-scale.

**Comment by matgis** [comment]
I think we respect the OS-level dpi scale factor on both Windows and macOS. 🤔

Shouldn't we do the same on Linux? It seems a bit strange to me to have logic like this in the application itself.

Or maybe it should be a preference?

How is it typically handled on Linux?

**Comment by ShalokShalom** [comment]
I have absolutely no idea, and I agree that it is probably possible to just accept the scaling factor of the system.

How is this handled for Windows and macOS, do you know the location of the code?

I imagine it might be because I am on KDE, and this is a command to scale GTK? 

I don't know, but should we implement this with DefOS?

**Comment by britzl** [comment]
> I don't know, but should we implement this with DefOS?

DefOS is only used in the runtime/engine.

> I imagine it might be because I am on KDE, and this is a command to scale GTK?

The editor uses JavaFX and the UI scaling and adapting to any OS wide UI scaling should happen there and not in the editor code. I suppose we need to research this a bit to understand if there's something we can do on our end. As far as I know the macOS and Windows versions of the editor respect the UI scale.

**Comment by ShalokShalom** [comment]
Where could we start looking?

**Comment by sprocketc** [comment]
Scaling seems to work fine on Wayland+Gnome.

@ShalokShalom Using `Dglass.gtk.uiScale` is the suggested solution according to https://wiki.archlinux.org/title/HiDPI#Java_applications

Setting `GDK_SCALE` might also work as a more generalized fix (e.g. `export GDK_SCALE=2`), although I am not sure if this supported in Wayland.

**Comment by ShalokShalom** [comment]
| The editor already scales well when I set vmargs = -Dglass.gtk.uiScale=200%, but I like it to auto-scale.

**Comment by sprocketc** [comment]
Yes, what you are currently doing might be the best option for the time being.

**Comment by ShalokShalom** [comment]
I am not asking for assistance. 

This is an issue report and feature request. 😄

**Comment by britzl** [comment]
> I agree that it is probably possible to just accept the scaling factor of the system.

How is the scaling factor configured on your Linux distribution? If there is a standardized way of configuring the entire OS level scale like on Window and macOS then that is something the Linux version of Defold should support (if it doesn't do this already).

On macOS there's this preference panel:

<img width="479" alt="Image" src="https://github.com/user-attachments/assets/6a140544-4d47-424e-908c-55bff1177ad3" />

And on Windows 11 there's something similar:

![Image](https://github.com/user-attachments/assets/7e73353e-a096-4aa1-a024-98feafd2781a)

What is the equivalent on popular Linux distros? Ubuntu has:

![Image](https://github.com/user-attachments/assets/b3a04002-3e4d-4bec-a9cb-4c8e4b6d4479)

Is this respected by Defold? Probably? If JavaFX doesn't pick up on some kind of OS level environment variable or setting and use that, then maybe we can do so in our launcher?

**Comment by ShalokShalom** [comment]
I do that on KDE, and its 175%

Wayland provides protocols for this:

`wp_fractional_scaling_v1`

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/143

And the client connects to an interface:

`wp_fractional_scale_manager`

**Comment by britzl** [comment]
After some internal discussion we've come to the conclusion that this should be handled in JavaFX and Defold should not do any workarounds or try to solve it on our own. There are multiple JavaFX issues related to this, for instance https://bugs.openjdk.org/browse/JDK-8160194?jql=status%20in%20(Open%2C%20%22In%20Progress%22%2C%20Reopened)%20AND%20text%20~%20%22HiDPI%20Linux%22

We will keep this issue open and test it if we see some progress or update JavaFX.

**Comment by ShalokShalom** [comment]
Well, I looked through these issues, and none of them seems to be worked on.
Also, the majority of them is many years old.
 
This one here seems to be the most relevant.
https://bugs.openjdk.org/browse/JDK-8326428

**Comment by rncar** [comment]
I use this on Arch Linux `GDK_SCALE=2 ./Defold`

**Comment by ShalokShalom** [comment]
This seems to be resolved. I installed the current version (defold-bin 1.10.2.5944edd-1 from **AUR**) and it works all fine on Arch Plasma Wayland. 

![Image](https://github.com/user-attachments/assets/eaa98162-9eba-49ca-a031-5d1e352b7a3c)
